functionary/prompt.py

```diff
@@ -31,147 +187,543 @@ def from_message(cls, message: Dict) -> EndToken:
+        response: Optional[Dict[str, Any]] = None
+        if current_state["response_type"] is None:
+            if cur_text.strip().startswith(self.start_function):  # if function_call
+                if cur_text.endswith(":"):
+                    f_index = cur_text.find(self.start_function)
+                    func_name = cur_text[
+                        f_index + len(self.start_function) : -1
+                    ].strip()
+                    response = {
+                        "delta": {
+                            "role": "assistant",
+                            "content": None,
+                            "function_call": {"arguments": "", "name": func_name},
+                        },
+                        "finish_reason": None,
+                    }
+                    current_state["response_type"] = "function"
+            else:  # if text_response
+                current_state["response_type"] = "text"
+                response = {
+                    "delta": {"content": "", "role": "assistant"},
+                    "finish_reason": None,
+                    "index": 0,
+                }
+
+        elif current_state["response_type"] == "function":
+            if finish_reason is None:
+                response = {
+                    "delta": {
+                        "role": "assistant",
+                        "function_call": {"arguments": delta_text},
+                    },  # format of openAI at the second return, don't need to add function_name
+                    "finish_reason": None,
+                    "index": 0,
+                }
+            else:
+                response = {
+                    "delta": {},
+                    "finish_reason": "function_call",
+                    "index": 0,
+                }  # format of openAI at the end, delta must be empty
+
+        elif current_state["response_type"] == "text":
+            if finish_reason is None:
+                # need to check if call a function or not
+                if cur_text.endswith(self.start_function):  # if call another function
+                    print("call another function in the mean time")
+                    cur_text = self.start_function
+                    current_state["current_text"] = self.start_function
+                    current_state["response_type"] = None
+                else:
+                    response = {
+                        "delta": {"content": delta_text, "role": "assistant"},
+                        "finish_reason": None,
+                        "index": 0,
+                    }
+            else:  # finish generating
+                response = {
+                    "delta": {},
+                    "finish_reason": finish_reason,
+                    "index": 0,
+                }  # format of openAI at the end, delta must be empty
+        return current_state, response
+
+    def get_chat_template(self) -> str:
+        chat_template = """{% for message in messages %}
```

Comment:
I think shouldn't because each prompt template has its own way of implementing jinja template, and why I prefer  to write a function instead of a fixed string because it would be easier to debug and fix and also easier to understand.

Reply:
I mean store chat_templates as variables in utils/chat_templates.py and import the respective chat template variable here.

Reply:
Actually, I think set it as a method would be easier to maintain. If you set it as a variable, it is like writting code without indents, and when you want to fix, it is really hard. You can also see that I manually add: "<b>" and then replace it to make it easier to write code.
By the way, we should put functions, variables in utils/chat_templates.py if they are supposed to be used multiple times, but for each prompt template, the jinja template is **unique** so I think we shouldn't

Reply:
Example for jinja template: 
```{% for message in messages %}
{% if message['role'] == 'user' or message['role'] == 'system' %}
{{ '<|from|>' + message['role'] + '
<|recipient|>all
<|content|>' + message['content'] + '
' }}{% elif message['role'] == 'tool' %}
{{ '<|from|>' + message['name'] + '
<|recipient|>all
<|content|>' + message['content'] + '
' }}{% else %}
{% set contain_content='no'%}
{% if message['content'] is not none %}
{{ '<|from|>assistant
<|recipient|>all
<|content|>' + message['content'] }}{% set contain_content='yes'%}
{% endif %}
{% if 'tool_calls' in message and message['tool_calls'] is not none %}
{% for tool_call in message['tool_calls'] %}
{% set prompt='<|from|>assistant
<|recipient|>' + tool_call['function']['name'] + '
<|content|>' + tool_call['function']['arguments'] %}
{% if loop.index == 1 and contain_content == "no" %}
{{ prompt }}{% else %}
{{ '
' + prompt}}{% endif %}
{% endfor %}
{% endif %}
{{ '<|stop|>
' }}{% endif %}
{% endfor %}
{% if add_generation_prompt %}{{ '<|from|>assistant
<|recipient|>' }}{% endif %}```

Reply:
Gotcha. Maybe we can implement importing while removing indents/whitespaces in future PR so it will be neater.